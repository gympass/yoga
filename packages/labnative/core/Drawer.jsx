import * as React from 'react';
import { createDrawerNavigator } from '@react-navigation/drawer';

import * as Pages from '../pages';
import NavigationDrawer from './NavigationDrawer';
import CenteredView from './CenteredView';

const Drawer = createDrawerNavigator();

function DrawerHeader(toggleDrawer) {
  return <NavigationDrawer toggleDrawer={toggleDrawer} />;
}

function DrawerComponent() {
  return (
    <Drawer.Navigator>
      {Object.keys(Pages).map(key => (
        <Drawer.Screen
          key={key}
          name={key}
          options={({ navigation: { toggleDrawer } }) => ({
            title: key,
            headerLeft: () => DrawerHeader(toggleDrawer),
          })}
        >
          {() => {
            const Page = Pages[key];

            return (
              <CenteredView>
                <Page />
              </CenteredView>
            );
          }}
        </Drawer.Screen>
      ))}
    </Drawer.Navigator>
  );
}

export default DrawerComponent;
