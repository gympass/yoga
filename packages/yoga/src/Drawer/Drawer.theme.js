const Drawer = ({ spacing }) => ({
  padding: {
    top: spacing.small,
    right: spacing.xxlarge,
    bottom: spacing.xxlarge,
    left: spacing.xxlarge,
  },
  width: {
    default: 600,
  },
});

export default Drawer;
