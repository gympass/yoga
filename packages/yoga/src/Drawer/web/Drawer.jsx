import React from 'react';
import styled from 'styled-components';
import { node, number } from 'prop-types';
import { theme } from '@gympass/yoga';
import Dialog from '../../Dialog';

const StyledDrawer = styled(Dialog)`
  position: absolute;
  right: 0;

  align-self: flex-end;

  height: 100%;
  padding: ${theme.spacing.zero} !important;
  border-radius: ${theme.borders.zero} !important;

  transition: 0.25s ease-in-out;

  animation: content;
  animation-duration: 400ms;
  animation-fill-mode: forwards;

  ${({
    theme: {
      yoga: {
        components: { drawer },
      },
    },
  }) => `
    width: min(${drawer.width.default}px, 100%);
  `}

  @keyframes content {
    0% {
      transform: translate3d(100%, 0, 0);
    }
  }
`;

const Drawer = React.forwardRef((props, forwardedRef) => (
  <StyledDrawer {...props} ref={forwardedRef} />
));

Drawer.propTypes = {
  children: node.isRequired,
  zIndex: number,
};

Drawer.defaultProps = {
  zIndex: 3,
};

export default Drawer;
