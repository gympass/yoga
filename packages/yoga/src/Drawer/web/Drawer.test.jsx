import React from 'react';
import { screen, render, cleanup } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ThemeProvider, Button } from '../..';

import Drawer from '.';

describe('<Drawer />', () => {
  afterEach(cleanup);

  it('should match snapshot', () => {
    const { baseElement } = render(
      <ThemeProvider>
        <Drawer isOpen>
          <Drawer.Header title="Title" />
          <Drawer.Content>Subtitle</Drawer.Content>
          <Drawer.Footer>
            <Button secondary>Ok, got it!</Button>
          </Drawer.Footer>
        </Drawer>
      </ThemeProvider>,
    );

    expect(baseElement).toMatchSnapshot();
  });

  it('should render a minimal drawer', async () => {
    const onActionMock = jest.fn();

    render(
      <ThemeProvider>
        <Drawer isOpen>
          <Drawer.Header title="Title" />
          <Drawer.Content>Subtitle</Drawer.Content>
          <Drawer.Footer>
            <Button onClick={onActionMock} secondary>
              Ok, got it!
            </Button>
          </Drawer.Footer>
        </Drawer>
      </ThemeProvider>,
    );

    const title = screen.getByText('Title');
    const subtitle = screen.getByText('Subtitle');

    const button = screen.getByRole('button', { name: /Ok, got it!/i });

    await userEvent.setup().click(button);

    expect(title).toBeVisible();
    expect(subtitle).toBeVisible();
    expect(onActionMock).toBeCalledTimes(1);
  });

  describe('<Drawer.Header />', () => {
    afterEach(cleanup);

    it('should render close bottom', () => {
      const onCloseActionMock = jest.fn();

      render(
        <ThemeProvider>
          <Drawer isOpen>
            <Drawer.Header title="Title" onClose={onCloseActionMock} />
          </Drawer>
        </ThemeProvider>,
      );

      const closeButton = screen.getByRole('button', {
        name: 'close-button-drawer',
      });

      expect(closeButton).toBeVisible();
    });

    it('should render back bottom', () => {
      const backHandleActionMock = jest.fn();

      render(
        <ThemeProvider>
          <Drawer isOpen>
            <Drawer.Header title="Title" backHandler={backHandleActionMock} />
          </Drawer>
        </ThemeProvider>,
      );

      const backButton = screen.getByRole('button', {
        name: 'back-button-drawer',
      });

      expect(backButton).toBeVisible();
    });

    it('should render close and back bottons', () => {
      const onCloseActionMock = jest.fn();
      const backHandleActionMock = jest.fn();

      render(
        <ThemeProvider>
          <Drawer isOpen>
            <Drawer.Header
              title="Title"
              onClose={onCloseActionMock}
              backHandler={backHandleActionMock}
            />
          </Drawer>
        </ThemeProvider>,
      );

      const backButton = screen.getByRole('button', {
        name: 'back-button-drawer',
      });
      const closeButton = screen.getByRole('button', {
        name: 'close-button-drawer',
      });

      expect(backButton).toBeVisible();
      expect(closeButton).toBeVisible();
    });

    it('should render divider component', () => {
      render(
        <ThemeProvider>
          <Drawer isOpen>
            <Drawer.Header title="Title" divider />
          </Drawer>
        </ThemeProvider>,
      );

      const divider = screen.getByRole('separator', { name: 'divider-drawer' });

      expect(divider).toBeInTheDocument();
    });

    it('should render drawer without drawer.header', async () => {
      const onButtonActionMock = jest.fn();

      render(
        <ThemeProvider>
          <Drawer isOpen>
            <Drawer.Content>Subtitle</Drawer.Content>
            <Drawer.Footer>
              <Button onClick={onButtonActionMock} secondary>
                Ok, got it!
              </Button>
            </Drawer.Footer>
          </Drawer>
        </ThemeProvider>,
      );

      const headerDrawer = screen.queryByRole('heading', {
        name: 'header-drawer',
      });

      const subtitle = screen.getByText('Subtitle');

      await userEvent
        .setup()
        .click(screen.getByRole('button', { name: /Ok, got it!/i }));

      expect(onButtonActionMock).toBeCalledTimes(1);

      expect(headerDrawer).not.toBeInTheDocument();
      expect(subtitle).toBeVisible();
    });

    it('should render onClose, backHandler and divider', async () => {
      const onCloseActionMock = jest.fn();
      const backHandleActionMock = jest.fn();

      render(
        <ThemeProvider>
          <Drawer isOpen>
            <Drawer.Header
              title="Title"
              onClose={onCloseActionMock}
              backHandler={backHandleActionMock}
              divider
            />
          </Drawer>
        </ThemeProvider>,
      );

      const closeButton = screen.getByRole('button', {
        name: 'close-button-drawer',
      });
      const backButton = screen.getByRole('button', {
        name: 'back-button-drawer',
      });

      const divider = screen.getByRole('separator', { name: 'divider-drawer' });

      expect(closeButton).toBeVisible();
      expect(backButton).toBeVisible();
      expect(divider).toBeVisible();
    });

    it('should call onClose function', async () => {
      const onCloseActionMock = jest.fn();

      render(
        <ThemeProvider>
          <Drawer isOpen>
            <Drawer.Header title="Title" onClose={onCloseActionMock} />
          </Drawer>
        </ThemeProvider>,
      );

      await userEvent
        .setup()
        .click(screen.getByRole('button', { name: 'close-button-drawer' }));

      expect(onCloseActionMock).toBeCalledTimes(1);
    });

    it('should call backHandler function', async () => {
      const backHandleActionMock = jest.fn();

      render(
        <ThemeProvider>
          <Drawer isOpen>
            <Drawer.Header title="Title" backHandler={backHandleActionMock} />
          </Drawer>
        </ThemeProvider>,
      );

      await userEvent
        .setup()
        .click(screen.getByRole('button', { name: 'back-button-drawer' }));

      expect(backHandleActionMock).toBeCalledTimes(1);
    });
  });
});
