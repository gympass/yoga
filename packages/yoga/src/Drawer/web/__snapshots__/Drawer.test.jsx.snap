// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> should match snapshot 1`] = `
.c14 {
  box-sizing: border-box;
  text-align: center;
  outline: none;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: 48px;
  padding-left: 24px;
  padding-right: 24px;
  background-color: #231B22;
  border: none;
  border-radius: 9999px;
  color: #FFFFFF;
  font-size: 16px;
  font-weight: 500;
  font-family: Rubik;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  line-height: 1;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c14 svg {
  width: 24px;
  height: 24px;
  fill: #FFFFFF;
  margin-right: 8px;
  -webkit-transition: fill 0.2s;
  transition: fill 0.2s;
}

.c14:not([disabled]):hover,
.c14:not([disabled]):focus {
  box-shadow: 0 4px 8px rgba(35,27,34,0.45);
}

.c14:active {
  background-color: rgba(35,27,34,0.75);
  color: #FFFFFF;
}

.c14:active svg {
  fill: #FFFFFF;
}

.c4 {
  width: 100%;
}

.c5 {
  padding-top: 16px;
  padding-right: 16px;
  padding-left: 40px;
}

.c8 {
  width: 100%;
  text-align: center;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c11 {
  margin-bottom: 32px;
  color: #6B6B78;
  font-size: 16px;
  font-weight: 400;
  text-align: center;
}

.c13 {
  padding-top: 16px;
  padding-right: 16px;
  padding-bottom: 40px;
  padding-left: 40px;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c9 {
  margin: 0;
  padding: 0;
  font-size: 16px;
  font-weight: 500;
  font-family: Rubik;
  color: #231B22;
  font-size: 14px;
}

.c1 {
  padding: 16px  16px  16px  16px;
  border-radius: 8px;
  background-color: #FFFFFF;
  box-shadow: 0 2px 4px -1px rgba(152,152,166,0.2),0 4px 5px 0px rgba(152,152,166,0.14),0 1px 10px 0px rgba(152,152,166,0.12);
}

.c2 {
  padding: 40px 32px 32px;
  width: 580px;
  min-height: 160px;
  border-radius: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  z-index: 3;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: fixed;
  top: 0;
  right: 0;
  left: 0;
  bottom: 0;
  background-color: rgba(35,27,34,0.48);
}

.c6 {
  background: #FFFFFF;
  padding: 16px 20px;
  width: 100%;
  min-height: 56px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0;
}

.c7 {
  width: 24px;
  height: 24px;
}

.c3 {
  position: absolute;
  right: 0;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  height: 100%;
  padding: 0 !important;
  border-radius: 0 !important;
  -webkit-transition: 0.25s ease-in-out;
  transition: 0.25s ease-in-out;
  -webkit-animation: content;
  animation: content;
  -webkit-animation-duration: 400ms;
  animation-duration: 400ms;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
  width: min(600px,100%);
}

.c12 {
  width: 100%;
  height: 100%;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  margin: 0;
  padding-left: 40px;
  text-align: left;
}

<body>
  <div />
  <div
    id="dialog"
  >
    <div>
      <div
        class="c0"
      >
        <section
          class="c1 c2 c3"
        >
          <div
            class="c4"
            width="100%"
          >
            <div
              class="c5"
            >
              <header
                aria-label="header-drawer"
                class="c6"
                role="heading"
              >
                <div
                  class="c7"
                />
                <div
                  class="c8"
                >
                  <h1
                    class="c9"
                    font-size="small"
                    style="letter-spacing: 1px; text-transform: uppercase; font-size: 12px;"
                  >
                    Title
                  </h1>
                </div>
                <div
                  class="c10"
                  display="flex"
                >
                  <div
                    class="c7"
                  />
                </div>
              </header>
            </div>
          </div>
          <div
            class="c11 c12"
            color="text.secondary"
          >
            Subtitle
          </div>
          <footer
            class="c13"
            d="flex"
            width="100%"
          >
            <button
              class="c14"
            >
              Ok, got it!
            </button>
          </footer>
        </section>
      </div>
    </div>
  </div>
</body>
`;
